Cython/Compiler/Code.py

```diff
@@ -1798,14 +1799,23 @@ def getvalue(self):
         return self.buffer.getvalue()
 
     def write(self, s):
-        # also put invalid markers (lineno 0), to indicate that those lines
-        # have no Cython source code correspondence
-        cython_lineno = self.last_marked_pos[1] if self.last_marked_pos else 0
+        # Cygdb needs to know which Cython source line corresponds to which C line.
+        # Therefore, we write the information which Cython line corresponds to which
+        # C line into self.buffer.markers and then write it from there into
+        # cython_debug/cython_debug_info_* (see ModuleNode._serialize_lineno_map).
+        # C lines with cython_lineno = 0 will get ingored by ModuleNode._serialize_lineno_map.
+        # The self.last_marked_pos[0] == self.source_desc check is necessary, because Cygdb
+        # assumes this to be true.
+        if self.last_marked_pos and self.last_marked_pos[0] == self.source_desc:
+            cython_lineno = self.last_marked_pos[1]
```

Comment:
Hmm … this doesn't feel great. I mean, we know the source file that we are currently generating code for. We write the name into the C file as a comment whenever a line starts. There doesn't seem to be a need to pass the source descriptor into the code writer's constructor, and in fact, it's even wrong, because it can generate code for an arbitrary number of source files (py/pyx, pxi, pxd) into the same C file.

I think this decision should solely depend on `last_marked_pos`.

Reply:
The reason why the old version didn't work is that if you would call `cy break mymodule:123` it would break once it hits line 123 in **any** cython file. I think the best solution would be to undo my chagnes and not only write self.last_marked_pos[1] into the file, but also self.last_marked_pos[0].

Reply:
Sounds like the right thing to do, yes.